runtime/descriptor.go

```diff
@@ -98,24 +105,66 @@ func propertySet(f *Frame, desc, inst, value *Object) *BaseException {
 
 // makeStructFieldDescriptor creates a descriptor with a getter that returns
 // the field given by fieldName from t's basis structure.
-func makeStructFieldDescriptor(t *Type, fieldName, propertyName string) *Object {
+func makeStructFieldDescriptor(t *Type, fieldName, propertyName string, fieldType fieldDescriptorType) *Object {
 	field, ok := t.basis.FieldByName(fieldName)
 	if !ok {
 		logFatal(fmt.Sprintf("no such field %q for basis %s", fieldName, nativeTypeName(t.basis)))
 	}
+
 	getterFunc := func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
-		var ret *Object
-		var raised *BaseException
-		if raised = checkFunctionArgs(f, fieldName, args, ObjectType); raised == nil {
-			o := args[0]
-			if !o.isInstance(t) {
+		if raised := checkFunctionArgs(f, fieldName, args, ObjectType); raised != nil {
+			return nil, raised
+		}
+
+		self := args[0]
+		if !self.isInstance(t) {
+			format := "descriptor '%s' for '%s' objects doesn't apply to '%s' objects"
+			return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), self.typ.Name()))
+		}
+
+		return WrapNative(f, t.slots.Basis.Fn(self).FieldByIndex(field.Index))
+	}
+
+	setter := None
+	if fieldType == fieldDescriptorRW {
+		if field.PkgPath != "" {
+			logFatal(fmt.Sprintf("field '%q' is not public on Golang code. Please fix it.", fieldName))
+		}
+
+		setterFunc := func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+			if raised := checkFunctionArgs(f, fieldName, args, ObjectType, ObjectType); raised != nil {
+				return nil, raised
+			}
+
+			self := args[0]
+			newValue := args[1]
+
+			if !self.isInstance(t) {
 				format := "descriptor '%s' for '%s' objects doesn't apply to '%s' objects"
-				raised = f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), o.typ.Name()))
-			} else {
-				ret, raised = WrapNative(f, t.slots.Basis.Fn(o).FieldByIndex(field.Index))
+				return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), self.typ.Name()))
+			}
+
+			if !newValue.isInstance(getNativeType(field.Type)) {
```

Comment:
The CPython traceback setting file.softspace to a string is: ``TypeError: an integer is required``.

For now, I am raising ``"an %s is required. Field '%s.%s' cannot store a '%s' value"``, just to be more helpful.

```python
# testing/file_test.py
f = open('/tmp/foo', 'w')
assert f.softspace == 0

f.softspace = 1
assert f.softspace == 1

try:
    f.softspace = 'deadbeef'
except TypeError as e:
    if 'is required' not in str(e):
        raise e     # Wrong exception arrived to us!
else:
    raise RuntimeError('a TypeError should had raised.')
```

Without this pre-check, Grumpy raises an ``TypeError: cannot convert string to int`` instead. 

And more intriguing: changing the sample code to ``f.softspace = '1'``, Grumpy still gives ``TypeError: cannot convert string to int``. CPython keeps on ``TypeError: an integer is required``.